### ISSUE TITLE
Missing stripslashes on referral description display

### ISSUE THREAD

**Issue by pippinsplugins** [AffiliateWP/AffiliateWP#1319]
![screen shot 2016-05-29 at 8 12 29 pm](https://cloud.githubusercontent.com/assets/1034109/15637240/b8635b42-25d9-11e6-9d6f-1a803606d794.png)

**Comment by pippinsplugins** [comment]
No, slashes are added on insert for escaping the data.

**Comment by TimothyBJacobs** [comment]
I think they are being added twice. Once in `wpdb::prepare` and once somewhere else. I believe the DB should have the unbackslashed version. Otherwise if you searched on the description field, for example, you'd get unexpected results.

**Comment by pippinsplugins** [comment]
Apostrophes in the database always need to be escaped to combat potential SQL injection. I don't believe the escape character actually affects querying on the value, especially since the values passed to the query would also be escaped if they contained single quotes, or other characters needing to be escaped.

**Comment by TimothyBJacobs** [comment]
Yes, but the escaping will leave the data clean in the database. For example.

```php
$wpdb->insert( 'wp_affiliate_wp_referrals', array(
        "affiliate_id" => 1,
        "visit_id" => 1,
        "description" => "Timothy's Referral"
), array("%d", "%d", "%s" ) );

var_dump( $wpdb->last_query );

$data = $wpdb->get_row( $wpdb->prepare("SELECT * FROM wp_affiliate_wp_referrals WHERE referral_id = %d", $wpdb->insert_id ) );
var_dump( $data->description );
```

Output:
```
string(120) "INSERT INTO `wp_affiliate_wp_referrals` (`affiliate_id`, `visit_id`, `description`) VALUES (1, 1, 'Timothy\'s Referral')"
string(18) "Timothy's Referral"
```

You would see a second backslash if the values were escaped a second time. For example, once by WordPress' magic quotes and once by `wpdb::prepare()`, or similar. That is why I think that they are being added a second time somewhere.

**Comment by pippinsplugins** [comment]
The data should definitely not be unslashed in the database in any scenario.

Are you seeing a double slash in the database somewhere?

**Comment by michaelbeil** [comment]
I am seeing a slash added on every save of a referral description: 
![screen capture on 2016-06-02 at 17-10-09](https://cloud.githubusercontent.com/assets/4752949/15762763/35bbd9ee-28e5-11e6-8c11-b6c6fde01e5b.gif)

**Comment by TimothyBJacobs** [comment]
I'm fairly certain that it should be.

Take for example `wp_insert_post()`. `wp_write_post()` is responsible for the creation of a new post and passes it the entire `$_POST` global. That data is slashed automatically by WordPress with `wp_magic_quotes()`.

`wp_insert_post()` does its processing, and then before insert calls `wp_unslash()` to remove those slashes before calling `wpdb::insert()`. This is to prevent `wpdb::prepare()` from escaping the quotes a second time. Escaping the quotes a second time would result in the backslash being persisted to the database.

You can verify this by creating a post in WordPress with single quotes. The data will be persisted without slashes.

**Comment by DrewAPicture** [comment]
@pippinsplugins I'm seeing inconsistent slashing behavior with the PR. Slashes are getting removed on display, but the text in the textarea is slashed in place when you edit it, and re-saving reslashes it. Sounds like we need to also strip slashes inside the editor textarea.

**Comment by pippinsplugins** [comment]
@DrewAPicture could you go ahead and adjust PR for that?

**Comment by TimothyBJacobs** [comment]
I'm going to reiterate that striping slashes on output doesn't make sense here and is the opposite of best practices to the best of my knowledge.

This can lead to unexpected behavior, for example.

```php
$referral = affwp_get_referral( 1 );
$referral->description .= " - And ain't this awesome.";

affiliate_wp()->referrals->update_referral( 1, [ 'description' => $referral->description ] );
```

The referral description is now stored with multiple slashes because the existing backslashes will be escaped, and the single quote that was added without magic quotes won't have any slashes in the database ( it will still be escaped during insert ).

Further, stripping slashes on output here means that slashes will have to be stripped on output everywhere, i.e. in WP-CLI, REST API, custom UIs, exporters, which is extremely prone to developer error.

As I said before, the data in the database should always be pristine. Slashing data has nothing to do with the storage of that data, but with the query to insert that data.

**Comment by pippinsplugins** [comment]
Unless slashes are stripped immediately after we retrieve the data (in our DB class), slashes should _always_ be stripped on output, never on input.

I'm sorry but it's simply not correct that data in the database should always be pristine. Data in the database should always be _safe_. Pristine and safe are very different. Safe in this case means slashed.

**Comment by TimothyBJacobs** [comment]
That's incorrect.

For example, from `wp_insert_post()`.

```php
$data = wp_unslash( $data );
$where = array( 'ID' => $post_ID );

if ( $update ) {
	if ( false === $wpdb->update( $wpdb->posts, $data, $where ) ) {
		if ( $wp_error ) {
			return new WP_Error('db_update_error', __('Could not update post in the database'), $wpdb->last_error);
		} else {
			return 0;
		}
	}
} else {
	if ( false === $wpdb->insert( $wpdb->posts, $data ) ) {
		if ( $wp_error ) {
			return new WP_Error('db_insert_error', __('Could not insert post into the database'), $wpdb->last_error);
		} else {
			return 0;
		}
	}
	$post_ID = (int) $wpdb->insert_id;
}
```

When retrieving a post via `get_post()`, the data is pulled directly from the database, and sanitized using `sanitize_post()` with an option of `raw`. This casts a few fields to integers and returns the data. No slashes are stripped. If you then access any `WP_Post` property, it will be unslashed.

You can verify that the data in the database is unslashed. Regardless of what role I used to create the post as well.

SQL injection prevention is handled automatically via `wpdb::prepare()` by iterating over all of the `$data` params and ( in addition to applying other things ), calling `mysqli_real_escape_string()`. The only time you need to provide your own escaping, through `esc_sql()`, is when building and executing a sql string manually without using `wpdb::prepare()`.

**Comment by pippinsplugins** [comment]
Ok this is making me question everything I've ever learned about slashed data . . . going to sit on it a bit.

**Comment by DrewAPicture** [comment]
Merged #1320
